pgmpy/prediction/NaivelVRegressor.py

```diff
@@ -0,0 +1,163 @@
+        stage1_estimator = clone(self.stage1_estimator)
+        stage2_estimator = clone(self.stage2_estimator)
+
+        exposure_vars = list(self.causal_graph.get_role("exposure"))
+        outcome_vars = list(self.causal_graph.get_role("outcome"))
+        instrument_vars = list(self.causal_graph.get_role("instrument"))
+
+        if len(exposure_vars) != 1:
+            raise ValueError(
+                f"NaiveIVRegressor requires exactly one exposure; got {len(exposure_vars)}"
+            )
+        if len(outcome_vars) != 1:
+            raise ValueError(
+                f"NaiveIVRegressor requires exactly one outcome; got {len(outcome_vars)}"
+            )
+        if len(instrument_vars) < 1:
+            raise ValueError(
+                f"NaiveIVRegressor requires at least one instrument; got {len(instrument_vars)}"
+            )
+
+        self.exposure_var_ = exposure_vars[0]
+        self.outcome_var_ = outcome_vars[0]
+        self.instrument_vars_ = instrument_vars
+        self.pretreatment_vars_ = list(self.causal_graph.get_role("pretreatment"))
+        self.feature_columns_ = (
+            [self.exposure_var_] + self.instrument_vars_ + self.pretreatment_vars_
+        )
+
+        # Handles sklearn n_features_in_ check by raising appropriate msg
+        X_arr = np.array(X)
+        n_features = 1 if X_arr.ndim == 1 else X_arr.shape[1]
+        self.n_features_in_ = n_features
+        # Handle sklearn single-feature check
+        if n_features == 1 and len(self.feature_columns_) > 1:
+            raise ValueError("1 feature(s) (n_features = 1)")
+
+        df = self._prepare_feature_df(X)
+
+        exposure_df = df[self.exposure_var_]
+        instrument_df = df[self.instrument_vars_]
+        pretreatment_df = df[self.pretreatment_vars_]
+
+        # fit stage1: E ~ Z
+        stage1_estimator.fit(instrument_df, exposure_df, sample_weight=sample_weight)
+        t_hat = stage1_estimator.predict(instrument_df)
+
+        # fit stage2: Y ~ t_hat + X
+        t_hat_2d = pd.DataFrame(t_hat.reshape(-1, 1))
+        covariates_df = pd.concat([t_hat_2d, pretreatment_df], axis=1)
+        stage2_estimator.fit(covariates_df, y, sample_weight=sample_weight)
+
+        # store
+        self.stage1_est_ = stage1_estimator
+        self.stage2_est_ = stage2_estimator
+
+        return self
+
+    def predict(self, X):
+        check_is_fitted(self, "stage1_est_")
+        check_is_fitted(self, "stage2_est_")
+
+        validate_data(
+            self, X, accept_sparse=False, ensure_2d=True, dtype="numeric", reset=False
+        )
+
+        X_df = self._prepare_feature_df(X)
+
+        instrument_df = X_df[self.instrument_vars_]
+        pre_treatment = X_df[self.pretreatment_vars_]
+
+        T_hat = self.stage1_est_.predict(instrument_df)
+        t_hat_2d = pd.DataFrame(T_hat.reshape(-1, 1))
+        covariates_df = pd.concat([t_hat_2d, pre_treatment], axis=1)
+        y_pred = self.stage2_est_.predict(covariates_df)
```

Comment:
I don't think the prediction algorithm is correct here. We only need to use the stage 2 estimator for making predictions. It would simply be `self.stage2_est_.predict(pd.concat([X, pre_treatment]))`. This is what was giving you issues with the `_prepare_features_df` method as the required columns for the `fit` and `predict` method are different. You could add a `required_columns` argument to `_prepare_feature_df` that is passed when the method is called and hence allows for different required columns in predict and fit. I am not sure if that would lead to some sklearn `check_estimator` checks to fail.